File: api/internal/admin/activity.go
package admin

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"

	"ooop-admin-api/internal/activity"
	"ooop-admin-api/internal/httpx"
)

// ===== 活动分类管理 =====

func (h *Handler) categoryList(c *gin.Context) {
	result, err := h.activities.AdminListCategories(c.Request.Context())
	writeResult(c, result, err)
}

func (h *Handler) categoryCreate(c *gin.Context) {
	var req struct {
		ID     string `json:"id"`
		Label  string `json:"label"`
		Icon   string `json:"icon"`
		Sort   int    `json:"sort"`
		Status int    `json:"status"`
	}
	if !bindJSON(c, &req) {
		return
	}
	result, err := h.activities.CreateCategory(c.Request.Context(), activity.CategoryInput{
		ID:     req.ID,
		Label:  req.Label,
		Icon:   req.Icon,
		Sort:   req.Sort,
		Status: req.Status,
	})
	writeResult(c, result, err)
}

func (h *Handler) categoryUpdate(c *gin.Context) {
	var req struct {
		Label  string `json:"label"`
		Icon   string `json:"icon"`
		Sort   int    `json:"sort"`
		Status int    `json:"status"`
	}
	if !bindJSON(c, &req) {
		return
	}
	result, err := h.activities.UpdateCategory(c.Request.Context(), c.Param("id"), activity.CategoryInput{
		Label:  req.Label,
		Icon:   req.Icon,
		Sort:   req.Sort,
		Status: req.Status,
	})
	writeResult(c, result, err)
}

func (h *Handler) categoryDelete(c *gin.Context) {
	id := c.Param("id")
	err := h.activities.DeleteCategory(c.Request.Context(), id)
	writeResult(c, gin.H{"id": id}, err)
}

// ===== 活动管理 =====

type activityUpdateRequest struct {
	Title             string `json:"title"`
	CategoryID        string `json:"category_id"`
	ActivityTime      string `json:"activity_time"`
	LocationText      string `json:"location_text"`
	City              string `json:"city"`
	TotalCount        int    `json:"total_count"`
	CostType          string `json:"cost_type"`
	FeeDetail         string `json:"fee_detail"`
	GenderRequirement string `json:"gender_requirement"`
	Intro             string `json:"intro"`
	Notice            string `json:"notice"`
}

func (r activityUpdateRequest) toUpdate() activity.AdminActivityUpdate {
	return activity.AdminActivityUpdate{
		Title:             r.Title,
		CategoryID:        r.CategoryID,
		ActivityTime:      r.ActivityTime,
		LocationText:      r.LocationText,
		City:              r.City,
		TotalCount:        r.TotalCount,
		CostType:          r.CostType,
		FeeDetail:         r.FeeDetail,
		GenderRequirement: r.GenderRequirement,
		Intro:             r.Intro,
		Notice:            r.Notice,
	}
}

func (h *Handler) activityList(c *gin.Context) {
	result, err := h.activities.AdminListActivities(c.Request.Context(), activity.AdminActivityQuery{
		Keyword:    c.Query("keyword"),
		Status:     c.Query("status"),
		CategoryID: c.Query("category_id"),
		Page:       queryInt(c, "page", 1),
		PageSize:   queryInt(c, "page_size", 10),
	})
	writeResult(c, result, err)
}

func (h *Handler) activityDetail(c *gin.Context) {
	id, ok := parseActivityID(c)
	if !ok {
		return
	}
	result, err := h.activities.GetActivityByID(c.Request.Context(), id)
	writeResult(c, result, err)
}

func (h *Handler) activityUpdate(c *gin.Context) {
	id, ok := parseActivityID(c)
	if !ok {
		return
	}
	var req activityUpdateRequest
	if !bindJSON(c, &req) {
		return
	}
	result, err := h.activities.UpdateActivity(c.Request.Context(), id, req.toUpdate())
	writeResult(c, result, err)
}

func (h *Handler) activityDelete(c *gin.Context) {
	id, ok := parseActivityID(c)
	if !ok {
		return
	}
	err := h.activities.DeleteActivity(c.Request.Context(), id)
	writeResult(c, gin.H{"id": id}, err)
}

func (h *Handler) activityReview(c *gin.Context) {
	id, ok := parseActivityID(c)
	if !ok {
		return
	}
	var req struct {
		Action string `json:"action"`
	}
	if !bindJSON(c, &req) {
		return
	}
	if req.Action != "approve" && req.Action != "reject" {
		httpx.Fail(c, http.StatusBadRequest, 400004, "审核动作不合法")
		return
	}
	result, err := h.activities.ReviewActivity(c.Request.Context(), id, req.Action == "approve")
	writeResult(c, result, err)
}

func (h *Handler) activityStatus(c *gin.Context) {
	id, ok := parseActivityID(c)
	if !ok {
		return
	}
	var req struct {
		Status string `json:"status"`
	}
	if !bindJSON(c, &req) {
		return
	}
	result, err := h.activities.SetActivityStatus(c.Request.Context(), id, req.Status)
	writeResult(c, result, err)
}

func parseActivityID(c *gin.Context) (int64, bool) {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil || id <= 0 {
		httpx.Fail(c, http.StatusBadRequest, 400001, "活动 ID 格式不正确")
		return 0, false
	}
	return id, true
}

File: web-admin/src/api/modules/activity.ts
import { get, post, put, del } from '@/utils/request'

export interface ActivityOrganizer {
  id: string
  name: string
}

// 后台活动项（取自后端 PublicActivity 的常用字段）。
export interface AdminActivity {
  id: string
  title: string
  categoryId: string
  categoryLabel: string
  imageUrl: string
  status: string
  costLabel: string
  costType: string
  timeRange: string
  activityTime: string
  dateText: string
  currentCount: number
  totalCount: number
  locationText: string
  city: string
  feeDetail: string
  genderRequirement: string
  intro: string
  notice: string
  organizer: ActivityOrganizer
  createdAt: string
}

export interface ActivityListParams {
  page: number
  page_size: number
  keyword?: string
  status?: string
  category_id?: string
}

export interface ActivityListResult {
  list: AdminActivity[]
  total: number
  page: number
  page_size: number
}

export interface PushChannelResult {
  channel: string
  triggered: boolean
  success: boolean
  message: string
  response?: string
}

export interface PushNotificationResult {
  triggered: boolean
  success: boolean
  alias: string
  message: string
  channels?: PushChannelResult[]
}

export interface AdminActivityReviewResult {
  activity: AdminActivity
  notification: PushNotificationResult
}

// 后台可改的活动文本字段（日期/坐标/图片/发起人/状态不在此处改）。
export interface UpdateActivityPayload {
  title: string
  category_id: string
  activity_time?: string
  location_text: string
  city: string
  total_count: number
  cost_type?: string
  fee_detail?: string
  gender_requirement?: string
  intro: string
  notice?: string
}

export function getActivityList(params: ActivityListParams) {
  return get<ActivityListResult>('admin/activities', { params })
}

export function getActivityDetail(id: string | number) {
  return get<AdminActivity>(`admin/activities/${id}`)
}

export function updateActivity(id: string | number, data: UpdateActivityPayload) {
  return put<AdminActivity>(`admin/activities/${id}`, data)
}

export function deleteActivity(id: string | number) {
  return del(`admin/activities/${id}`)
}

// 审核：通过 / 拒绝（仅对待审核活动生效）。
export function reviewActivity(id: string | number, action: 'approve' | 'reject') {
  return put<AdminActivityReviewResult>(`admin/activities/${id}/review`, { action })
}

// 上下架：taken_down 下架 / ongoing 恢复。
export function setActivityStatus(id: string | number, status: 'taken_down' | 'ongoing') {
  return put<AdminActivity>(`admin/activities/${id}/status`, { status })
}

// ===== 活动分类 =====

export interface AdminCategory {
  id: string
  label: string
  icon: string
  sort: number
  status: number
}

export interface CategoryPayload {
  id?: string
  label: string
  icon: string
  sort: number
  status: number
}

export function getCategoryList() {
  return get<AdminCategory[]>('admin/activity-categories')
}

export function createCategory(data: CategoryPayload) {
  return post<AdminCategory>('admin/activity-categories', data)
}

export function updateCategory(id: string, data: CategoryPayload) {
  return put<AdminCategory>(`admin/activity-categories/${id}`, data)
}

export function deleteCategory(id: string) {
  return del(`admin/activity-categories/${id}`)
}
